File: perl/calculator.pl
use strict;
use warnings;

my $ERR_NONE = 0;
my $ERR_INVALID_EXPR = 1;
my $ERR_DIV_BY_ZERO = 2;

sub parse_expression {
    my ($expr) = @_;
    $expr =~ s/\s+//g; 
    
    if ($expr eq '') {
        return (0, $ERR_INVALID_EXPR, "Boş ifade");
    }
    
    if ($expr =~ /^[0-9]+$/) {
        return (int($expr), $ERR_NONE, "");
    }
    
    if ($expr =~ /^\((.+)\)$/) {
        return parse_expression($1);
    }
    
    my $depth = 0;
    my $lowest_precedence = 999;
    my $lowest_op_pos = -1;
    
    my @chars = split //, $expr;
    for (my $i = 0; $i < scalar @chars; $i++) {
        my $c = $chars[$i];
        if ($c eq '(') {
            $depth++;
        } elsif ($c eq ')') {
            if ($depth > 0) {
                $depth--;
            } else {
                return (0, $ERR_INVALID_EXPR, "Parantezler eşleşmiyor");
            }
        } elsif (($c eq '+' || $c eq '-') && $depth == 0) {
            if ($i == 0 || $chars[$i-1] !~ /[0-9\)]/) {
                next;
            }
            
            if ($lowest_precedence >= 1) {
                $lowest_precedence = 1;
                $lowest_op_pos = $i;
            }
        } elsif (($c eq '*' || $c eq '/') && $depth == 0) {
            if ($lowest_precedence >= 2) {
                $lowest_precedence = 2;
                $lowest_op_pos = $i;
            }
        }
    }
    
    if ($depth != 0) {
        return (0, $ERR_INVALID_EXPR, "Parantezler eşleşmiyor");
    }
    
    if ($lowest_op_pos >= 0) {
        my $left_expr = substr($expr, 0, $lowest_op_pos);
        my $right_expr = substr($expr, $lowest_op_pos + 1);
        my $op = $chars[$lowest_op_pos];
        
        my ($left, $left_err, $left_err_msg) = parse_expression($left_expr);
        if ($left_err != $ERR_NONE) {
            return ($left, $left_err, $left_err_msg);
        }
        
        my ($right, $right_err, $right_err_msg) = parse_expression($right_expr);
        if ($right_err != $ERR_NONE) {
            return ($right, $right_err, $right_err_msg);
        }
        
        if ($op eq '+') {
            return (int($left + $right), $ERR_NONE, "");
        } elsif ($op eq '-') {
            return (int($left - $right), $ERR_NONE, "");
        } elsif ($op eq '*') {
            return (int($left * $right), $ERR_NONE, "");
        } elsif ($op eq '/') {
            if ($right == 0) {
                return (0, $ERR_DIV_BY_ZERO, "Sıfıra bölme hatası");
            }
            return (int($left / $right), $ERR_NONE, "");
        }
    }
    
    return (0, $ERR_INVALID_EXPR, "İfade anlaşılamadı: $expr");
}

print "Çıkmak için 'exit' yazın\n";

while (1) {
    print "> ";
    my $input = <STDIN>;
    chomp($input);
    
    last if $input =~ /^\s*$/;
    
    if ($input eq "exit" || $input eq "quit") {
        last;
    } else {
        my ($result, $error, $error_msg) = parse_expression($input);
        
        if ($error != $ERR_NONE) {
            print "Hata: $error_msg\n";
        } else {
            print "= $result\n";
        }
    }
} 

File: rust/calculator.rs
use std::io;

enum CalculatorError {
    InvalidExpression(String),
    DivisionByZero
}

struct Calculator {}

impl Calculator {
    fn new() -> Calculator {
        Calculator {}
    }
    
    fn evaluate(&self, expression: &str) -> Result<f64, CalculatorError> {
        self.parse_expression(expression)
    }
    
    fn parse_expression(&self, expression: &str) -> Result<f64, CalculatorError> {
        let expression = expression.trim();
        
        if expression.is_empty() {
            return Err(CalculatorError::InvalidExpression("Boş ifade".to_string()));
        }
        
        let expression: String = expression.chars().filter(|c| !c.is_whitespace()).collect();
        
        if !expression.contains('+') && !expression.contains('-') && 
           !expression.contains('*') && !expression.contains('/') && 
           !expression.contains('(') && !expression.contains(')') {
            
            if let Ok(number) = expression.parse::<f64>() {
                return Ok(number);
            }
            
            return Err(CalculatorError::InvalidExpression(format!("Geçersiz sayı: {}", expression)));
        }
        
        if expression.starts_with('(') && expression.ends_with(')') {
            return self.parse_expression(&expression[1..expression.len()-1]);
        }
        
        let mut operator_pos = None;
        let mut parenthesis_level = 0;
        
        for (i, c) in expression.chars().enumerate() {
            match c {
                '(' => parenthesis_level += 1,
                ')' => parenthesis_level -= 1,
                '+' | '-' => {
                    if parenthesis_level == 0 && i > 0 {
                        let prev = expression.chars().nth(i-1).unwrap();
                        if prev.is_digit(10) || prev == ')' {
                            operator_pos = Some(i);
                            break;
                        }
                    }
                },
                _ => {}
            }
        }
        
        if operator_pos.is_none() {
            parenthesis_level = 0;
            
            for (i, c) in expression.chars().enumerate() {
                match c {
                    '(' => parenthesis_level += 1,
                    ')' => parenthesis_level -= 1,
                    '*' | '/' => {
                        if parenthesis_level == 0 {
                            operator_pos = Some(i);
                            break;
                        }
                    },
                    _ => {}
                }
            }
        }
        
        if let Some(pos) = operator_pos {
            let left_expr = &expression[0..pos];
            let left_value = self.parse_expression(left_expr)?;
            
            let right_expr = &expression[pos+1..];
            let right_value = self.parse_expression(right_expr)?;
            
            match expression.chars().nth(pos).unwrap() {
                '+' => return Ok(left_value + right_value),
                '-' => return Ok(left_value - right_value),
                '*' => return Ok(left_value * right_value),
                '/' => {
                    if right_value == 0.0 {
                        return Err(CalculatorError::DivisionByZero);
                    }
                    return Ok(left_value / right_value);
                },
                _ => return Err(CalculatorError::InvalidExpression("Bilinmeyen operatör".to_string()))
            }
        }
        
        Err(CalculatorError::InvalidExpression(format!("İfade anlaşılamadı: {}", expression)))
    }
}

fn main() {
    let calculator = Calculator::new();
    println!("Çıkmak için 'exit' yazın");
    
    loop {
        print!("-> ");
      
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Girdi okunamadı");
        
        let input = input.trim();
        
        if input.is_empty() {
            continue;
        }
        
        match input {
            "exit" => break,
            _ => {
                match calculator.evaluate(input) {
                    Ok(result) => println!("= {}", result),
                    Err(error) => {
                        match error {
                            CalculatorError::InvalidExpression(message) => println!("Hata: {}", message),
                            CalculatorError::DivisionByZero => println!("Hata: Sıfıra bölme işlemi yapılamaz"),
                        }
                    }
                }
            }
        }
    }
}